src/main/java/at/ac/tuwien/kr/alpha/common/program/AbstractProgram.java

```diff
@@ -0,0 +1,52 @@
+package at.ac.tuwien.kr.alpha.common.program;
+
+import java.util.Collections;
+import java.util.List;
+
+import at.ac.tuwien.kr.alpha.Util;
+import at.ac.tuwien.kr.alpha.common.atoms.Atom;
+import at.ac.tuwien.kr.alpha.common.rule.AbstractRule;
+import at.ac.tuwien.kr.alpha.common.rule.head.Head;
+import at.ac.tuwien.kr.alpha.grounder.parser.InlineDirectives;
+
+/**
+ * The parent type for all kinds of program. Defines a program's basic structure (facts + rules + inlineDirectives)
+ * 
+ * @param <R> the type of rule a program permits. This needs to be determined by implementations based on which syntax constructs an implementation permits
+ *        Copyright (c) 2019, the Alpha Team.
+ */
+public abstract class AbstractProgram<R extends AbstractRule<? extends Head>> {
```

Comment:
Your suggestion would result in a compiler warning - since `AbstractRule` is a generic type, `<R extends AbstractRule>` would be compiled to something along the lines of `<R extends AbstractRule<Object>>`.

The idea is that an `AbstractProgram` can consist of any type of `AbstractRule`. Typically, the respective `Rule` implementation would be a non-generic type, i.e. bind the type parameter from `AbstractRule`, but it's theoretically possible to use generic rule types as well. (That's just a general thought though, the real reason for the type signature is - as explained above - that it's the correct way to declare a generic type parameter that is itself a generic)